File: Data Structures & Algorithms/valid-parenthesis-string/submission-2.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        n=len(s)
        dp={}
        

      

        def backtrack(i:int,openCount:int)->bool:
            if (i,openCount) in dp:
                return dp[(i,openCount)]
            if i==n:
                dp[(i,openCount)]= openCount==0
                return dp[(i,openCount)]
        
            if s[i]=='(':
            
                dp[(i,openCount)]= backtrack(i+1,openCount+1)
                return dp[(i,openCount)]
  
            elif s[i]==')':
                if not openCount:
                    dp[(i,openCount)]=False
                    return dp[(i,openCount)]
          
                dp[(i,openCount)]= backtrack(i+1,openCount-1)
                return dp[(i,openCount)]
            else:
                if openCount:
                    dp[(i,openCount)]= backtrack(i+1,openCount-1) or backtrack(i+1,openCount+1) or backtrack(i+1,openCount)
                else:
                    dp[(i,openCount)]=  backtrack(i+1,openCount+1) or backtrack(i+1,openCount)
                return dp[(i,openCount)]
        return backtrack(0,0)
                





            

        
        return True




                

        

File: Data Structures & Algorithms/n-queens/submission-2.java
class Solution {
    public void backtrack(int n,int i,HashSet<Integer> cols,HashSet<Integer> lrDiag,HashSet<Integer> rlDiag,List<StringBuilder> curr,List<List<String>> res){
        if(i==n){
            List<String> currRes=new ArrayList<>();
            for(int j=0;j<n;j++){
                currRes.add(curr.get(j).toString());
            }

            res.add(currRes);
            return;
        }
        for(int j=0;j<n;j++){
            if((!cols.contains(j))&&(!lrDiag.contains(i+j))&&(!rlDiag.contains(i-j))){
                cols.add(j);
                lrDiag.add(i+j);
                rlDiag.add(i-j);
                curr.get(i).setCharAt(j,'Q');
                backtrack(n,i+1,cols,lrDiag,rlDiag,curr,res);
                cols.remove(j);
                lrDiag.remove(i+j);
                rlDiag.remove(i-j);
                curr.get(i).setCharAt(j,'.');

            }
        }

    }
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res=new ArrayList<>();
        List<StringBuilder> curr=new ArrayList<>();
        for(int i=0;i<n;i++){
            curr.add(new StringBuilder(".".repeat(n)));
        }

        HashSet<Integer> cols=new HashSet<>();
        HashSet<Integer> lrDiag=new HashSet<>();
        HashSet<Integer> rlDiag=new HashSet<>();

        backtrack(n,0,cols,lrDiag,rlDiag,curr,res);
        return res;

      
        

  






        
    }
}
